### PR TITLE
goto-instrument: replace "modifies" by "assigns" terminology

### LINKED ISSUE #6467: Rename "modifies*" entities in order to avoid possible confusion
CBMC version: 516f109218a522bf33ade14e3e20003c1427beb0
Operating system: n/a
Exact command line resulting in the issue: n/a (internal code refactor)
What behaviour did you expect: n/a (internal code refactor)
What happened instead: n/a (internal code refactor)

Related to @tautschnig's comment on a recent PR: https://github.com/diffblue/cbmc/pull/6249#discussion_r751149114

There are several methods and classes in goto-instrument that implement "assigns" (i.e., "writes") semantics but are named "modifies*" (e.g., `get_modifies` method in `loop_utils.h`, `function_modifiest` class in `function_modifies.h` etc.) We should refactor these and rename them to "assigns*" to avoid any possible confusion in the future. `x := x` is a non-modifying write.

The `function_modifiest` member function implementations also look quite similar to methods within `loop_utils.cpp`. We should also look into unifying these implementations.

### PR DESCRIPTION
**Pull request by tautschnig**
All this code collects assignments rather than checking for genuine value changes (modifications).

Fixes: #6467

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by TGWDB — APPROVED** [review]

**Review by peterschrammel — APPROVED** [review]

File: src/goto-instrument/function_assigns.cpp [hunk 1]
```diff
@@ -0,0 +1,78 @@
+/*******************************************************************\
+
+Module: assigns
```
Comment (feliperodri) [inline src/goto-instrument/function_assigns.cpp]: Why not `function_assigns`?

File: src/goto-instrument/function_assigns.cpp [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+Author: Daniel Kroening, kroening@kroening.com
+
+\*******************************************************************/
+
+/// \file
+/// assigns
```
Comment (feliperodri) [inline src/goto-instrument/function_assigns.cpp]: Why not `function_assigns.cpp`?

File: src/goto-instrument/function_assigns.h [hunk 3]
```diff
@@ -0,0 +1,51 @@
+/*******************************************************************\
+
+Module: assigns
```
Comment (feliperodri) [inline src/goto-instrument/function_assigns.h]: Why not `function_assigns`?

File: src/goto-instrument/function_assigns.h [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+Author: Daniel Kroening, kroening@kroening.com
+
+\*******************************************************************/
+
+/// \file
+/// assigns
```
Comment (feliperodri) [inline src/goto-instrument/function_assigns.h]: Why not `function_assigns.h`?

**Review by feliperodri — APPROVED** [review]

**Review by feliperodri — APPROVED** [review]

**Pull request merged**